File: src/omnia/data/structure/rx/ObservableUndirectedGraph.java
package omnia.data.structure.rx;

import io.reactivex.Flowable;
import omnia.data.structure.Set;
import omnia.data.structure.UndirectedGraph;

public interface ObservableUndirectedGraph<E> {

  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableGraph.ObservableChannels<E> {

    @Override
    Flowable<? extends UndirectedGraph<E>> states();

    @Override
    Flowable<? extends MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableGraph.MutationEvent<E> {

    @Override
    UndirectedGraph<E> state();

    @Override
    Set<? extends ObservableGraph.GraphOperation<E>> operations();
  }
}

File: omnia/src/commonMain/omnia/data/structure/observable/ObservableUndirectedGraph.kt
package omnia.data.structure.observable

import com.badoo.reaktive.observable.Observable
import omnia.data.structure.Set
import omnia.data.structure.UndirectedGraph
import omnia.data.structure.observable.ObservableGraph.GraphOperation

interface ObservableUndirectedGraph<E : Any> : ObservableGraph<E>, UndirectedGraph<E> {

  override val observables: Observables<E>

  interface Observables<E : Any> : ObservableGraph.Observables<E> {

    override val states: Observable<UndirectedGraph<E>>

    override val mutations: Observable<MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableGraph.MutationEvent<E> {

    override val state: UndirectedGraph<E>

    override val operations: Set<out GraphOperation<E>>
  }
}